packages/core/src/modules/mdoc/MdocDeviceResponse.ts

```diff
@@ -0,0 +1,193 @@
+  }
+
+  private static createPresentationSubmission(input: {
+    id: string
+    presentationDefinition: {
+      id: string
+      input_descriptors: ReturnType<typeof MdocDeviceResponse.assertMdocInputDescriptor>[]
+    }
+  }) {
+    const { id, presentationDefinition } = input
+    if (presentationDefinition.input_descriptors.length !== 1) {
+      throw new MdocError('Currently Mdoc Presentation Submissions can only be created for a sigle input descriptor')
+    }
+    return {
+      id,
+      definition_id: presentationDefinition.id,
+      descriptor_map: [
+        {
+          id: presentationDefinition.input_descriptors[0].id,
+          format: 'mso_mdoc',
+          path: '$',
+        },
+      ],
+    }
+  }
+
+  public static limitDisclosureToInputDescriptor(options: { inputDescriptor: InputDescriptorV2; mdoc: Mdoc }) {
+    const { mdoc } = options
+
+    const inputDescriptor = this.assertMdocInputDescriptor(options.inputDescriptor)
+    const _mdoc = parseIssuerSigned(TypedArrayEncoder.fromBase64(mdoc.base64Url), mdoc.docType)
+    return mdocLimitDisclosureToId({ mdoc: _mdoc, inputDescriptor })
+  }
+
+  public static async openId4Vp(agentContext: AgentContext, options: MdocDeviceResponseOpenId4VpOptions) {
+    const { sessionTranscriptOptions } = options
+    const presentationDefinition = this.partitionPresentationDefinition(
+      options.presentationDefinition
+    ).mdocPresentationDefinition
+
+    const issuerSignedDocuments = options.mdocs.map((mdoc) =>
+      parseIssuerSigned(TypedArrayEncoder.fromBase64(mdoc.base64Url), mdoc.docType)
+    )
+    const mdoc = new MDoc(issuerSignedDocuments)
+
+    // TODO: we need to implement this differently.
+    // TODO: Multiple Mdocs can have different device keys.
+    const mso = mdoc.documents[0].issuerSigned.issuerAuth.decodedPayload
+    const deviceKeyInfo = mso.deviceKeyInfo
+    if (!deviceKeyInfo?.deviceKey) {
+      throw new CredoError('Device key info is missing')
+    }
+
+    const publicDeviceJwk = COSEKey.import(deviceKeyInfo.deviceKey).toJWK()
+
+    deviceKeyInfo.deviceKey
+    const deviceResponseBuilder = await DeviceResponse.from(mdoc)
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      .usingPresentationDefinition(presentationDefinition as any)
+      .usingSessionTranscriptForOID4VP(sessionTranscriptOptions)
+      .authenticateWithSignature(publicDeviceJwk, 'ES256')
+
+    for (const [nameSpace, nameSpaceValue] of Object.entries(options.deviceNameSpaces ?? {})) {
+      deviceResponseBuilder.addDeviceNameSpace(nameSpace, nameSpaceValue)
+    }
+
+    const deviceResponseMdoc = await deviceResponseBuilder.sign(getMdocContext(agentContext))
+
+    return {
+      deviceResponseBase64Url: TypedArrayEncoder.toBase64URL(deviceResponseMdoc.encode()),
+      presentationSubmission: MdocDeviceResponse.createPresentationSubmission({
+        id: 'MdocPresentationSubmission ' + agentContext.wallet.generateNonce(),
```

Comment:
I think we can also just use an uuid?

```suggestion
        id: 'MdocPresentationSubmission ' + await agentContext.wallet.generateNonce(),
```